File: photon-targeting/src/main/java/org/photonvision/estimation/VisionEstimation.java
/*
 * Copyright (C) Photon Vision.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.photonvision.estimation;

import edu.wpi.first.apriltag.AprilTag;
import edu.wpi.first.apriltag.AprilTagFieldLayout;
import edu.wpi.first.math.MatBuilder;
import edu.wpi.first.math.Matrix;
import edu.wpi.first.math.Nat;
import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Transform2d;
import edu.wpi.first.math.geometry.Transform3d;
import edu.wpi.first.math.geometry.Translation3d;
import edu.wpi.first.math.numbers.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import org.ejml.simple.SimpleMatrix;
import org.opencv.calib3d.Calib3d;
import org.opencv.core.MatOfDouble;
import org.opencv.core.MatOfPoint2f;
import org.opencv.core.Point;
import org.photonvision.jni.ConstrainedSolvepnpJni;
import org.photonvision.targeting.PhotonTrackedTarget;
import org.photonvision.targeting.PnpResult;
import org.photonvision.targeting.TargetCorner;

public class VisionEstimation {
    /** Get the visible {@link AprilTag}s which are in the tag layout using the visible tag IDs. */
    public static List<AprilTag> getVisibleLayoutTags(
            List<PhotonTrackedTarget> visTags, AprilTagFieldLayout tagLayout) {
        return visTags.stream()
                .map(
                        t -> {
                            int id = t.getFiducialId();
                            var maybePose = tagLayout.getTagPose(id);
                            return maybePose.map(pose3d -> new AprilTag(id, pose3d)).orElse(null);
                        })
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    /**
     * Performs solvePNP using 3d-2d point correspondences of visible AprilTags to estimate the
     * field-to-camera transformation. If only one tag is visible, the result may have an alternate
     * solution.
     *
     * <p><b>Note:</b> The returned transformation is from the field origin to the camera pose!
     *
     * <p>With only one tag: {@link OpenCVHelp#solvePNP_SQUARE}
     *
     * <p>With multiple tags: {@link OpenCVHelp#solvePNP_SQPNP}
     *
     * @param cameraMatrix The camera intrinsics matrix in standard opencv form
     * @param distCoeffs The camera distortion matrix in standard opencv form
     * @param visTags The visible tags reported by PV. Non-tag targets are automatically excluded.
     * @param tagLayout The known tag layout on the field
     * @return The transformation that maps the field origin to the camera pose. Ensure the {@link
     *     PnpResult} are present before utilizing them.
     */
    public static Optional<PnpResult> estimateCamPosePNP(
            Matrix<N3, N3> cameraMatrix,
            Matrix<N8, N1> distCoeffs,
            List<PhotonTrackedTarget> visTags,
            AprilTagFieldLayout tagLayout,
            TargetModel tagModel) {
        if (tagLayout == null
                || visTags == null
                || tagLayout.getTags().isEmpty()
                || visTags.isEmpty()) {
            return Optional.empty();
        }

        var corners = new ArrayList<TargetCorner>();
        var knownTags = new ArrayList<AprilTag>();
        // ensure these are AprilTags in our layout
        for (var tgt : visTags) {
            int id = tgt.getFiducialId();
            tagLayout
                    .getTagPose(id)
                    .ifPresent(
                            pose -> {
                                knownTags.add(new AprilTag(id, pose));
                                corners.addAll(tgt.getDetectedCorners());
                            });
        }
        if (knownTags.isEmpty() || corners.isEmpty() || corners.size() % 4 != 0) {
            return Optional.empty();
        }
        Point[] points = OpenCVHelp.cornersToPoints(corners);

        // single-tag pnp
        if (knownTags.size() == 1) {
            var camToTag =
                    OpenCVHelp.solvePNP_SQUARE(cameraMatrix, distCoeffs, tagModel.vertices, points);
            if (!camToTag.isPresent()) return Optional.empty();
            var bestPose = knownTags.get(0).pose.transformBy(camToTag.get().best.inverse());
            var altPose = new Pose3d();
            if (camToTag.get().ambiguity != 0)
                altPose = knownTags.get(0).pose.transformBy(camToTag.get().alt.inverse());

            var o = new Pose3d();
            return Optional.of(
                    new PnpResult(
                            new Transform3d(o, bestPose),
                            new Transform3d(o, altPose),
                            camToTag.get().ambiguity,
                            camToTag.get().bestReprojErr,
                            camToTag.get().altReprojErr));
        }
        // multi-tag pnp
        else {
            var objectTrls = new ArrayList<Translation3d>();
            for (var tag : knownTags) objectTrls.addAll(tagModel.getFieldVertices(tag.pose));
            var camToOrigin = OpenCVHelp.solvePNP_SQPNP(cameraMatrix, distCoeffs, objectTrls, points);
            if (camToOrigin.isEmpty()) return Optional.empty();
            return Optional.of(
                    new PnpResult(
                            camToOrigin.get().best.inverse(),
                            camToOrigin.get().alt.inverse(),
                            camToOrigin.get().ambiguity,
                            camToOrigin.get().bestReprojErr,
                            camToOrigin.get().altReprojErr));
        }
    }

    /**
     * Performs constrained solvePNP using 3d-2d point correspondences of visible AprilTags to
     * estimate the field-to-camera transformation.
     *
     * <p><b>Note:</b> The returned transformation is from the field origin to the robot drivebase
     *
     * @param cameraMatrix The camera intrinsics matrix in standard opencv form
     * @param distCoeffs The camera distortion matrix in standard opencv form
     * @param visTags The visible tags reported by PV. Non-tag targets are automatically excluded.
     * @param robot2camera The {@link Transform3d} from the robot odometry frame to the camera optical
     *     frame
     * @param robotPoseSeed An initial guess at robot pose, refined via optimizaiton. Better guesses
     *     will converge faster.
     * @param tagLayout The known tag layout on the field
     * @param tagModel The physical size of the AprilTags
     * @param headingFree If heading is completely free, or if our measured gyroθ is taken into
     *     account
     * @param gyroθ If headingFree is false, the best estimate at robot yaw. Excursions from this are
     *     penalized in our cost function.
     * @param gyroErrorScaleFac A relative weight for gyro heading excursions against tag corner
     *     reprojection error.
     * @return The transformation that maps the field origin to the camera pose. Ensure the {@link
     *     PnpResult} are present before utilizing them.
     */
    public static Optional<PnpResult> estimateRobotPoseConstrainedSolvepnp(
            Matrix<N3, N3> cameraMatrix,
            Matrix<N8, N1> distCoeffs,
            List<PhotonTrackedTarget> visTags,
            Transform3d robot2camera,
            Pose3d robotPoseSeed,
            AprilTagFieldLayout tagLayout,
            TargetModel tagModel,
            boolean headingFree,
            Rotation2d gyroθ,
            double gyroErrorScaleFac) {
        if (tagLayout == null
                || visTags == null
                || tagLayout.getTags().isEmpty()
                || visTags.isEmpty()) {
            return Optional.empty();
        }

        var corners = new ArrayList<TargetCorner>();
        var knownTags = new ArrayList<AprilTag>();
        // ensure these are AprilTags in our layout
        for (var tgt : visTags) {
            int id = tgt.getFiducialId();
            tagLayout
                    .getTagPose(id)
                    .ifPresent(
                            pose -> {
                                knownTags.add(new AprilTag(id, pose));
                                corners.addAll(tgt.getDetectedCorners());
                            });
        }
        if (knownTags.isEmpty() || corners.isEmpty() || corners.size() % 4 != 0) {
            return Optional.empty();
        }
        Point[] points = OpenCVHelp.cornersToPoints(corners);

        // Undistort
        {
            MatOfPoint2f temp = new MatOfPoint2f();
            MatOfDouble cameraMatrixMat = new MatOfDouble();
            MatOfDouble distCoeffsMat = new MatOfDouble();
            OpenCVHelp.matrixToMat(cameraMatrix.getStorage()).assignTo(cameraMatrixMat);
            OpenCVHelp.matrixToMat(distCoeffs.getStorage()).assignTo(distCoeffsMat);

            temp.fromArray(points);
            Calib3d.undistortImagePoints(temp, temp, cameraMatrixMat, distCoeffsMat);
            points = temp.toArray();

            temp.release();
            cameraMatrixMat.release();
            distCoeffsMat.release();
        }

        // Rotate from wpilib to opencv camera CS
        var robot2cameraBase =
                MatBuilder.fill(Nat.N4(), Nat.N4(), 0, 0, 1, 0, -1, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0, 1);
        var robotToCamera = robot2camera.toMatrix().times(robot2cameraBase);

        // Where we saw corners
        var point_observations = new SimpleMatrix(2, points.length);
        for (int i = 0; i < points.length; i++) {
            point_observations.set(0, i, points[i].x);
            point_observations.set(1, i, points[i].y);
        }

        // Affine corner locations
        var objectTrls = new ArrayList<Translation3d>();
        for (var tag : knownTags) objectTrls.addAll(tagModel.getFieldVertices(tag.pose));
        var field2points = new SimpleMatrix(4, points.length);
        for (int i = 0; i < objectTrls.size(); i++) {
            field2points.set(0, i, objectTrls.get(i).getX());
            field2points.set(1, i, objectTrls.get(i).getY());
            field2points.set(2, i, objectTrls.get(i).getZ());
            field2points.set(3, i, 1);
        }

        // fx fy cx cy
        double[] cameraCal =
                new double[] {
                    cameraMatrix.get(0, 0),
                    cameraMatrix.get(1, 1),
                    cameraMatrix.get(0, 2),
                    cameraMatrix.get(1, 2),
                };

        var guess2 = robotPoseSeed.toPose2d();

        var ret =
                ConstrainedSolvepnpJni.do_optimization(
                        headingFree,
                        knownTags.size(),
                        cameraCal,
                        robotToCamera.getData(),
                        new double[] {
                            guess2.getX(), guess2.getY(), guess2.getRotation().getRadians(),
                        },
                        field2points.getDDRM().getData(),
                        point_observations.getDDRM().getData(),
                        gyroθ.getRadians(),
                        gyroErrorScaleFac);

        if (ret == null) {
            return Optional.empty();
        } else {
            var pnpresult = new PnpResult();
            pnpresult.best = new Transform3d(new Transform2d(ret[0], ret[1], new Rotation2d(ret[2])));
            return Optional.of(pnpresult);
        }
    }
}

File: photon-targeting/src/main/native/cpp/photon/estimation/VisionEstimation.cpp
/*
 * Copyright (C) Photon Vision.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "photon/estimation/VisionEstimation.h"

#include <iostream>
#include <utility>
#include <vector>

#include "photon/constrained_solvepnp/wrap/casadi_wrapper.h"
#include "photon/estimation/OpenCVHelp.h"
#include "photon/targeting/MultiTargetPNPResult.h"

namespace photon {
namespace VisionEstimation {

std::vector<frc::AprilTag> GetVisibleLayoutTags(
    const std::vector<PhotonTrackedTarget>& visTags,
    const frc::AprilTagFieldLayout& layout) {
  std::vector<frc::AprilTag> retVal{};
  for (const auto& tag : visTags) {
    int id = tag.GetFiducialId();
    auto maybePose = layout.GetTagPose(id);
    if (maybePose) {
      retVal.emplace_back(frc::AprilTag{id, maybePose.value()});
    }
  }
  return retVal;
}

std::optional<PnpResult> EstimateCamPosePNP(
    const Eigen::Matrix<double, 3, 3>& cameraMatrix,
    const Eigen::Matrix<double, 8, 1>& distCoeffs,
    const std::vector<PhotonTrackedTarget>& visTags,
    const frc::AprilTagFieldLayout& layout, const TargetModel& tagModel) {
  if (visTags.size() == 0) {
    return PnpResult();
  }

  std::vector<photon::TargetCorner> corners{};
  std::vector<frc::AprilTag> knownTags{};

  for (const auto& tgt : visTags) {
    int id = tgt.GetFiducialId();
    auto maybePose = layout.GetTagPose(id);
    if (maybePose) {
      knownTags.emplace_back(frc::AprilTag{id, maybePose.value()});
      auto currentCorners = tgt.GetDetectedCorners();
      corners.insert(corners.end(), currentCorners.begin(),
                     currentCorners.end());
    }
  }
  if (knownTags.size() == 0 || corners.size() == 0 || corners.size() % 4 != 0) {
    return PnpResult{};
  }

  std::vector<cv::Point2f> points = OpenCVHelp::CornersToPoints(corners);

  if (knownTags.size() == 1) {
    auto camToTag = OpenCVHelp::SolvePNP_Square(cameraMatrix, distCoeffs,
                                                tagModel.GetVertices(), points);
    if (!camToTag) {
      return PnpResult{};
    }
    frc::Pose3d bestPose =
        knownTags[0].pose.TransformBy(camToTag->best.Inverse());
    frc::Pose3d altPose{};
    if (camToTag->ambiguity != 0) {
      altPose = knownTags[0].pose.TransformBy(camToTag->alt.Inverse());
    }
    frc::Pose3d o{};
    PnpResult result{};
    result.best = frc::Transform3d{o, bestPose};
    result.alt = frc::Transform3d{o, altPose};
    result.ambiguity = camToTag->ambiguity;
    result.bestReprojErr = camToTag->bestReprojErr;
    result.altReprojErr = camToTag->altReprojErr;
    return result;
  } else {
    std::vector<frc::Translation3d> objectTrls{};
    for (const auto& tag : knownTags) {
      auto verts = tagModel.GetFieldVertices(tag.pose);
      objectTrls.insert(objectTrls.end(), verts.begin(), verts.end());
    }
    auto ret = OpenCVHelp::SolvePNP_SQPNP(cameraMatrix, distCoeffs, objectTrls,
                                          points);
    if (ret) {
      // Invert best/alt transforms
      ret->best = ret->best.Inverse();
      ret->alt = ret->alt.Inverse();
    }

    return ret;
  }
}

std::optional<photon::PnpResult> EstimateRobotPoseConstrainedSolvePNP(
    const Eigen::Matrix<double, 3, 3>& cameraMatrix,
    const Eigen::Matrix<double, 8, 1>& distCoeffs,
    const std::vector<photon::PhotonTrackedTarget>& visTags,
    const frc::Transform3d& robot2Camera, const frc::Pose3d& robotPoseSeed,
    const frc::AprilTagFieldLayout& layout, const photon::TargetModel& tagModel,
    bool headingFree, frc::Rotation2d gyroTheta, double gyroErrorScaleFac) {
  if (visTags.size() == 0) {
    return photon::PnpResult();
  }

  std::vector<photon::TargetCorner> corners{};
  std::vector<frc::AprilTag> knownTags{};

  for (const auto& tgt : visTags) {
    int id = tgt.GetFiducialId();
    auto maybePose = layout.GetTagPose(id);
    if (maybePose) {
      knownTags.emplace_back(frc::AprilTag{id, maybePose.value()});
      auto currentCorners = tgt.GetDetectedCorners();
      corners.insert(corners.end(), currentCorners.begin(),
                     currentCorners.end());
    }
  }
  if (knownTags.size() == 0 || corners.size() == 0 || corners.size() % 4 != 0) {
    return photon::PnpResult{};
  }

  std::vector<cv::Point2f> points =
      photon::OpenCVHelp::CornersToPoints(corners);

  cv::Mat cameraMat(cameraMatrix.rows(), cameraMatrix.cols(), CV_64F);
  cv::eigen2cv(cameraMatrix, cameraMat);
  cv::Mat distCoeffsMat(distCoeffs.rows(), distCoeffs.cols(), CV_64F);
  cv::eigen2cv(distCoeffs, distCoeffsMat);

  cv::undistortImagePoints(points, points, cameraMat, distCoeffsMat);

  Eigen::Matrix4d robotToCameraBase{
      (Eigen::Matrix4d() << 0, 0, 1, 0, -1, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0, 1)
          .finished()};
  Eigen::Matrix<constrained_solvepnp::casadi_real, 4, 4, Eigen::ColMajor>
      robotToCamera{robot2Camera.ToMatrix() * robotToCameraBase};

  Eigen::Matrix<constrained_solvepnp::casadi_real, 2, Eigen::Dynamic,
                Eigen::ColMajor>
      pointObservations{};
  pointObservations.resize(2, points.size());
  for (size_t i = 0; i < points.size(); i++) {
    pointObservations(0, i) = points[i].x;
    pointObservations(1, i) = points[i].y;
  }

  std::vector<frc::Translation3d> objectTrls{};
  for (const auto& tag : knownTags) {
    auto verts = tagModel.GetFieldVertices(tag.pose);
    objectTrls.insert(objectTrls.end(), verts.begin(), verts.end());
  }

  Eigen::Matrix<constrained_solvepnp::casadi_real, 4, Eigen::Dynamic,
                Eigen::ColMajor>
      field2points{};
  field2points.resize(4, objectTrls.size());
  for (size_t i = 0; i < objectTrls.size(); i++) {
    field2points(0, i) = objectTrls[i].X().value();
    field2points(1, i) = objectTrls[i].Y().value();
    field2points(2, i) = objectTrls[i].Z().value();
    field2points(3, i) = 1;
  }

  frc::Pose2d guess2 = robotPoseSeed.ToPose2d();

  constrained_solvepnp::CameraCalibration cameraCal{
      cameraMatrix(0, 0),
      cameraMatrix(1, 1),
      cameraMatrix(0, 2),
      cameraMatrix(1, 2),
  };

  Eigen::Matrix<constrained_solvepnp::casadi_real, 3, 1> guessMat{
      guess2.X().value(), guess2.Y().value(),
      guess2.Rotation().Radians().value()};

  wpi::expected<constrained_solvepnp::RobotStateMat, slp::ExitStatus> result =
      constrained_solvepnp::do_optimization(
          headingFree, knownTags.size(), cameraCal, robotToCamera, guessMat,
          field2points, pointObservations, gyroTheta.Radians().value(),
          gyroErrorScaleFac);

  if (!result.has_value()) {
    return {};
  } else {
    photon::PnpResult res{};

    res.best = frc::Transform3d{frc::Transform2d{
        units::meter_t{result.value()[0]}, units::meter_t{result.value()[1]},
        frc::Rotation2d{units::radian_t{result.value()[2]}}}};

    return res;
  }
}

}  // namespace VisionEstimation
}  // namespace photon
